### PR TITLE
[v3] Implement RFC 7366 - Encrypt-then-MAC

### PR DESCRIPTION
**Pull request by tomato42**
Implements the encrypt then MAC mechanism for CBC based cipher suites in TLS. Add some simple tests for HandshakeSettings.

Missing:
- client side check for EtM advertised in server hello together with stream or AEAD cipher (it's incorrect use and it will fail to interoperate but won't report correct errors - I'd rather test it using mechanism introduced in #47)
- ability to force use of EtM (abort connection when server didn't select EtM but did select CBC cipher suite - that will need to wait for AEAD to be actually useful)

Changes since v2:
- reworked "move MAC-then-encrypt", "etm decryption" and "etm decryption" to fix most pylint complaints
- rebased on master

Changes since v1:
- reworked "etm decryption" and "etm encryption" patches to better show what is happening, unfortunately the "move MAC-then-encrypt to separate methods" is still rather hard to read, but it's not really possible to make it any better
- rebased on master with fixed tests

Obsoletes #52

### CODE REVIEW

**Comment by tomato42** [comment]
when can I expect review?

**Comment by trevp** [comment]
Next week.

File: tlslite/handshakesettings.py [hunk 1]
```diff
@@ -106,10 +106,17 @@ def __init__(self):
         self.minVersion = (3,1)
         self.maxVersion = (3,3)
         self.useExperimentalTackExtension = False
+        self.useEncryptThenMAC = True
 
-    # Validates the min/max fields, and certificateTypes
-    # Filters out unsupported cipherNames and cipherImplementations
-    def _filter(self):
+    def validate(self):
```
Comment (trevp) [inline tlslite/handshakesettings.py]: this rename could be in a separate pull request

File: tlslite/handshakesettings.py [hunk 2]
```diff
@@ -144,6 +152,8 @@ def _filter(self):
             raise ValueError("maxKeySize too small")
         if other.maxKeySize>16384:
             raise ValueError("maxKeySize too large")
+        if other.maxKeySize < other.minKeySize:
```
Comment (trevp) [inline tlslite/handshakesettings.py]: this could also be separate

File: tlslite/tlsconnection.py [hunk 3]
```diff
@@ -399,15 +399,23 @@ def _handshakeClientAsyncHelper(self, srpParams, certParams, anonParams,
         #We'll use this for the ClientHello, and if an error occurs
         #parsing the Server Hello, we'll use this version for the response
         self.version = settings.maxVersion
+
+        if settings.useEncryptThenMAC:
+            extensions = [TLSExtension().create(ExtensionType.encrypt_then_mac,
+                                                bytearray(0))]
+        else:
+            extensions = None
```
Comment (trevp) [inline tlslite/tlsconnection.py]: I think it would be cleaner to create the Extensions list within clientSendClientHello(), instead of creating it here and passing it in.

File: tlslite/tlsconnection.py [hunk 4]
```diff
@@ -399,15 +399,23 @@ def _handshakeClientAsyncHelper(self, srpParams, certParams, anonParams,
...
         
         # OK Start sending messages!
         # *****************************
 
         # Send the ClientHello.
         for result in self._clientSendClientHello(settings, session, 
-                                        srpUsername, srpParams, certParams,
-                                        anonParams, serverName, nextProtos,
-                                        reqTack):
+                                                  srpUsername, srpParams,
```
Comment (trevp) [inline tlslite/tlsconnection.py]: see previous comment.

File: tlslite/tlsconnection.py [hunk 5]
```diff
@@ -423,6 +431,10 @@ def _handshakeClientAsyncHelper(self, srpParams, certParams, anonParams,
         # (string or None)
         nextProto = self._clientSelectNextProto(nextProtos, serverHello)
 
+        # check if server supports encrypt_then_mac
+        if serverHello.getExtension(ExtensionType.encrypt_then_mac):
+            self.etm = True
```
Comment (trevp) [inline tlslite/tlsconnection.py]: let's rename etm -> encryptThenMac

File: tlslite/tlsconnection.py [hunk 6]
```diff
@@ -1145,10 +1160,22 @@ def _handshakeServerAsyncHelper(self, verifierDB,
             tackExt = TackExtension.create(tacks, activationFlags)
         else:
             tackExt = None
+
+        # prepare encrypt then mac if required
+        if settings.useEncryptThenMAC and\
+                clientHello.getExtension(ExtensionType.encrypt_then_mac) and\
+                not cipherSuite in [CipherSuite.TLS_RSA_WITH_RC4_128_SHA,\
```
Comment (trevp) [inline tlslite/tlsconnection.py]: Checking against RC4 like this is ugly, maybe add a special suite for it?

File: tlslite/tlsrecordlayer.py [hunk 7]
```diff
@@ -148,6 +153,9 @@ def __init__(self, sock):
         #Fault we will induce, for testing purposes
         self.fault = None
 
+        #Whatever to do Encrypt and MAC or MAC and Encrypt
+        self.etm = False
```
Comment (trevp) [inline tlslite/tlsrecordlayer.py]: "Whether" to do..., also rename etm -> encryptThenMac

File: tlslite/tlsrecordlayer.py [hunk 8]
```diff
@@ -951,89 +1021,173 @@ def _getNextRecord(self):
...
-            macLength = self._readState.macContext.digest_size
-            endLength = macLength + totalPaddingLength
-            if endLength > len(b):
-                macGood = False
+        # Remove padding
+        b = b[:-totalPaddingLength]
+
+        yield b
+
+    def _decryptRecordWithMtE(self, recordType, b):
```
Comment (trevp) [inline tlslite/tlsrecordlayer.py]: don't like seeing the padding and decryption code duplicated, could this be refactored, for example a single decryptRecord() function?

File: unit_tests/test_tlslite_handshakesettings.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
+# Author: Hubert Kario (c) 2014
+# see LICENCE file for legal information regarding use of this file
```
Comment (trevp) [inline unit_tests/test_tlslite_handshakesettings.py]: This could be handled in a separate pull request.

**Comment by trevp** [comment]
Hi Hubert, added some line comments.  Also:
 - have you done interop testing w/another library?
 - RFC 7366 mentions that once EtM has been negotiated, it should be used for resumed sessions, could you add that?

**Comment by tomato42** [comment]
> have you done interop testing w/another library?
yes, I tested against the eid server mentioned in drafts
 > RFC 7366 mentions that once EtM has been negotiated, it should be used for resumed sessions, could you add that?
I'll look into that

File: tlslite/tlsrecordlayer.py (hunk 8 above)
Comment (tomato42) [inline tlslite/tlsrecordlayer.py]: while application of padding can be moved outside, the decryption code needs to stay separate as the handling of errors is completely different in EtM and MtE case

**Comment by tomato42** [comment]
fully obsoleted, closing
